include/sms.h

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#ifndef SMS_H_
+#define SMS_H_
+
+/**
+ * @file sms.h
+ *
+ * @defgroup sms SMS subscriber manager
+ *
+ * @{
+ *
+ * @brief Public APIs of the SMS subscriber manager module.
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <zephyr/types.h>
+#include <sys/types.h>
+
+/* Forward declaration. */
+struct sms_data;
+
+/** @brief SMS listener callback function. */
+typedef void (*sms_callback_t)(struct sms_data *const data, void *context);
+
+/**
+ * @brief SMS subscriber handle.
+ * A negative value indicates an invalid handle.
+ */
+typedef ssize_t sms_handle_t;
```

Comment:
int<num>_t seems more natural to use here since this is an identifier, and not a size.

Reply:
I would avoid the `typedef` and use `int`, you can call the param `sms_handle` in the API.

Reply:
This handle identifies the (handler, context) pair. This is why I would like to use a new type. Is this allowed?
My implementation uses the index of an array as the handle, this is why I choose to use `size_t`. Any `int_t` size could work as well. It will just limit the number of subscribers.

Reply:
My two cents,

it's not disallowed to add new types, it's discouraged though - that's why you get the following warning: `WARNING:NEW_TYPEDEFS: do not add new typedefs`.

My personal opinion is that covering simple types with custom typedefs brings more bad than good - it does not provide any extra type protection (you can mixup `sms_handle_t`/`int`/`ssize_t` however you want, compiler won't complain), yet it makes the code harder to analyze (as you always need to check what exactly the type is whenever you encounter a new type). So in this case, I'd back @lemrey opinion.

Reply:
> My implementation uses the index of an array as the handle, this is why I choose to use size_t. Any int_t size could work as well. It will just limit the number of subscribers.

My understanding is that size_t should be used for sizes of things, an index into an array is not a size and should therefore not be used. It's not clear to me how this will limit the number of subscribers.

Reply:
My understanding was wrong, using size_t for indexes is fine.

Reply:
I like to use `typedef` to keep the API binary compatible.
I will use and `int` as we do in the `bsdlib` to represent handles.